File: typescript/src/problems101_200/best-time-to-buy-and-sell-stock2.ts
class Problem {
  maxProfit(prices: number[]): number {
    throw new Error(`Not implemented (${prices}).`);
  }
}

class Default extends Problem {
  maxProfit(prices: number[]): number {
    let profit = 0;
    let hold = false;
    for (let i = 0; i < prices.length; i++) {
      if (i !== prices.length - 1 && prices[i + 1] > prices[i]) {
        if (!hold) {
          hold = true;
          profit -= prices[i];
        }
        continue;
      }
      if (!hold) {
        continue;
      }
      hold = false;
      profit += prices[i];
    }
    return profit;
  }
}

const SOLUTIONS = [new Default()];
export default SOLUTIONS;

File: python/src/problems101_200/best_time_to_buy_and_sell_stock2.py
from abc import ABC, abstractmethod
from typing import override


class Problem(ABC):
    @abstractmethod
    def max_profit(self, prices: list[int]) -> int:
        pass


class Default(Problem):
    @override
    def max_profit(self, prices: list[int]) -> int:
        profit: int = 0
        hold: bool = False
        for i, price in enumerate(prices):
            if i != len(prices) - 1 and prices[i + 1] > price:
                if not hold:
                    hold = True
                    profit -= prices[i]
                continue
            if not hold:
                continue
            hold = False
            profit += prices[i]
        return profit


SOLUTIONS: list[Problem] = [Default()]
